### ISSUE TITLE
Inferring the original type from the complex type instance

### ISSUE THREAD

**Issue by dfee** [mobxjs/mobx-state-tree#1211]
**_Bug report_**

* [x] I've checked documentation and searched for existing issues
* [x] I've made sure my project is based on the latest MST version
* [x] Fork [this](https://codesandbox.io/s/y64pzxj01) code sandbox or another minimal reproduction. 

**Sandbox link or minimal reproduction code**
```ts
const MyStore = types.model("MyStore", {
  foo: types.number,
});

type MyStoreType = typeof MyStore;
type MyStoreInstance = Instance<MyStoreType>;

type InferredTypeFromInstance<T> = T extends Instance<infer U> ? U : never;
type InferredMyStoreType = InferredTypeFromInstance<MyStoreInstance>;

function doesExtend<T extends U, U>() {};
doesExtend<InferredMyStoreType, MyStoreType>(); // type error: does not extend!
doesExtend<InferredMyStoreType, MyStoreInstance>(); // does extend! ?!?!
```

**Describe the expected behavior**
The inferred type from `Instance<infer U>` should be the model's type.

**Describe the observed behavior**
The inferred type from `Instance<infer U>` is the supplied InstanceType, `T`.

**Comment by xaviergonz** [comment]
The thing is that `Instance<some MST type>` will give the instance type, but currently the `Instance<not a MST type>` will give that "not a MST type"

This means:
```
type InferredTypeFromInstance<T> = T extends Instance<infer U> ? U : never; // this will always be "true"
type InferredMyStoreType = InferredTypeFromInstance<MyStoreInstance>; // this will return the instance itself
```

Even if Instance was "fixed" to return never for non MST types though (so something like `type Instance2<T> = T extends IType<any, any, infer TT> ? TT : never`), that still wouldn't work though:
```
    type InferredTypeFromInstance<T> = T extends Instance2<infer U> ? U : never;
    type InferredMyStoreType = InferredTypeFromInstance<MyStoreInstance>; // this would be never
```
since the test would never be true since Instances do not have in their template arguments a reference back to the original type. For that to work the instance would need to save the original type somewhere.

Imagine this:
```
    type TFirstArg<T> = T extends [infer A, any] ? A : T
    type MyT = [number, string]
    type MyT2 = TFirstArg<MyT> // this is number
    // but you can't get back from MyT2 (number) to MyT ([number, string]) since the data is "lost" (not saved in MyT2)
```

What are you trying to achieve?

**Comment by dfee** [comment]
I am trying to create a method on an abstract "service" class that, based on a property name in the root store, performs an operation returning an instance of that service.

for some more concrete details, here was my investigation on [spectrum.chat yesterday](https://spectrum.chat/?t=467ec87c-f81a-4bae-a1b3-2b8f2bada702).

**Comment by xaviergonz** [comment]
After https://github.com/mobxjs/mobx-state-tree/pull/1229 is merged there will be a `TypeOfValue<typeof VARIABLE>` that will return the original type of complex values

**Comment by xaviergonz** [comment]
Released as 3.12.0, @dfee let us know how it goes :)
